Code/GraphMol/Chirality.cpp

```diff
@@ -451,65 +451,141 @@ bool atomIsCandidateForRingStereochem(const ROMol &mol, const Atom *atom) {
   return res;
 }
 
-// returns true if the atom is allowed to have stereochemistry specified
-bool checkChiralAtomSpecialCases(ROMol &mol, const Atom *atom) {
-  PRECONDITION(atom, "bad atom");
-
+// finds all possible chiral special cases.
+// at the moment this is just candidates for ring stereochemistry
+void findChiralAtomSpecialCases(ROMol &mol,
+                                boost::dynamic_bitset<> &possibleSpecialCases) {
+  PRECONDITION(possibleSpecialCases.size() >= mol.getNumAtoms(),
+               "bit vector too small");
+  possibleSpecialCases.reset();
   if (!mol.getRingInfo()->isInitialized()) {
     VECT_INT_VECT sssrs;
     MolOps::symmetrizeSSSR(mol, sssrs);
   }
-
-  const RingInfo *ringInfo = mol.getRingInfo();
-  if (ringInfo->numAtomRings(atom->getIdx()) &&
-      atomIsCandidateForRingStereochem(mol, atom)) {
-    // the atom is in a ring, so the "chirality" specification may actually
-    // be handling ring stereochemistry.
-
-    // check for another chiral tagged
-    // atom without stereochem in this atom's rings:
+  boost::dynamic_bitset<> atomsSeen(mol.getNumAtoms());
+  boost::dynamic_bitset<> atomsUsed(mol.getNumAtoms());
+  boost::dynamic_bitset<> bondsSeen(mol.getNumBonds());
+
+  for (ROMol::AtomIterator ait = mol.beginAtoms(); ait != mol.endAtoms();
+       ++ait) {
+    const Atom *atom = *ait;
+    if (atomsSeen[atom->getIdx()]) continue;
+    if (atom->getChiralTag() == Atom::CHI_UNSPECIFIED ||
+        atom->hasProp(common_properties::_CIPCode) ||
```

Comment:
The reason ring stereochemistry is hard (in the current representation) is that the atoms don't have CIP chirality but do have (and need to have) the stereochemistry specified. That's what this is testing for: the stereochemistry is specified and there's no CIP code on the atom.